pkg/dev/dev.go

```diff
@@ -74,3 +95,109 @@ func (o *DevClient) Cleanup() error {
 	var err error
 	return err
 }
+
+// SetupPortForwarding sets up port forwarding for the endpoints in the devfile
+func (o *DevClient) SetupPortForwarding(devfileObj parser.DevfileObj, envSpecificInfo *envinfo.EnvSpecificInfo, out, errOut io.Writer) error {
+	var err error
+	var containers []v1.Component
+	containers, err = devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1.ContainerComponentType},
+	})
+
+	endpoints := endpointsFromContainers(containers)
+	portPairs := portPairsFromEndpoints(endpoints)
+
+	if len(portPairs) == 0 {
+		// no endpoints with exposure set to public or internal; no ports to be forwarded
+		return nil
+	}
+
+	var pod *corev1.Pod
+	pod, err = o.kubernetesClient.GetOnePodFromSelector(componentlabels.GetSelector(devfileObj.GetMetadataName(), envSpecificInfo.GetApplication()))
+	if err != nil {
+		return err
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(o.kubernetesClient.GetClientConfig())
+	if err != nil {
+		return err
+	}
+
+	req := o.kubernetesClient.GeneratePortForwardReq(pod.Name)
+
+	var dialer httpstream.Dialer
+	dialer = spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+
+	stopChan := make(chan struct{}, 1)
+	// passing nil for readyChan because it's eventually being closed if it's not nil
+	// passing nil for out because we only care for error, not for output messages; we want to print our own messages
+	var fw *portforward.PortForwarder
+	fw, err = portforward.NewOnAddresses(dialer, []string{"localhost"}, portPairs, stopChan, nil, nil, errOut)
+	if err != nil {
+		return err
+	}
+
+	// start port-forwarding
+	go func() {
+		err = fw.ForwardPorts()
+		if err != nil {
+			fmt.Fprintf(out, fmt.Errorf("error setting up port forwarding: %v", err).Error())
+			os.Exit(1)
```

Comment:
Any suggestions for doing this better? Especially the `fmt.Fprintf` line makes me cringe. The `os.Exit(1)` line seems OK to me because, IMO, we should exit if port-forwarding fails. But open to suggestions.

Reply:
`os.Exit(1)` is going to cause problems when you try to implement #5485 
If this fails the cleanup needs to happen. `os.Exit` will prevent you from doing that.

Reply:
Yeah, that's one thing that is on my mind. But what else could be done here? Should we just notify the user on CLI and not stop the `odo dev` execution? That doesn't seem correct behaviour either. But I'm fine by changing it to not exit.

Reply:
It would also mess with the telemetry, I think. Telemetry won't have a way to know why the command stopped. Correct me if I am wrong.

Reply:
@kadel @feloy I have added a TODO to tackle this with the cleanup task.